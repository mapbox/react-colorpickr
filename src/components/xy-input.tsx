import React, { useRef, ReactNode } from 'react';
import themeable from 'react-themeable';
import { autokey } from '../autokey';
import clamp from 'clamp';

interface Props {
  children: ReactNode;
  theme: { [id: string]: string };
  x: number;
  y: number;
  xmax: number;
  ymax: number;
  isDark: boolean;
  discRadius: number;
  onChange: ({ x, y }: { x: number; y: number }) => void;
  backgroundColor: string;
}

function XYInput({
  children,
  theme,
  x,
  y,
  xmax,
  ymax,
  isDark,
  discRadius,
  onChange,
  backgroundColor
}: Props) {
  const xyControlContainer = useRef(null);
  const coords = useRef({ start: { x: 0, y: 0 }, offset: { x: 0, y: 0 } });
  const themer = autokey(themeable(theme));
  const top = Math.round(clamp((y / ymax) * 100, 0, 100));
  const left = Math.round(clamp((x / xmax) * 100, 0, 100));

  const change = ({ top, left }: { top: number; left: number }) => {
    const { width, height } =
      xyControlContainer.current.getBoundingClientRect();

    onChange({
      x: (clamp(left, 0, width) / width) * xmax,
      y: (clamp(top, 0, height) / height) * ymax
    });
  };

  const dragEnd = (e: MouseEvent) => {
    e.preventDefault();
    document.removeEventListener('mousemove', drag);
    document.removeEventListener('touchmove', drag, {
      passive: false
    } as unknown as EventListenerOptions);
    document.removeEventListener('mouseup', dragEnd);
    document.removeEventListener('touchend', dragEnd);
    document.removeEventListener('touchcancel', dragEnd);
  };

  const drag = (e: MouseEvent | TouchEvent) => {
    const { start, offset } = coords.current;
    e.preventDefault();

    let x: number;
    let y: number;

    if (e instanceof TouchEvent) {
      x = e.changedTouches[0].clientX;
      y = e.changedTouches[0].clientY;
    } else {
      x = e.clientX;
      y = e.clientY;
    }

    const top = y + start.y - offset.y;
    const left = x + start.x - offset.x;
    change({ top, left });
  };

  const dragStart = (e: MouseEvent | TouchEvent) => {
    e.preventDefault();

    const rect = xyControlContainer.current.getBoundingClientRect();
    let x: number;
    let y: number;

    if (e instanceof TouchEvent) {
      x = e.changedTouches[0].clientX;
      y = e.changedTouches[0].clientY;
    } else {
      x = e.clientX;
      y = e.clientY;
    }

    const offset = {
      left: x - rect.left,
      top: y - rect.top
    };

    change(offset);

    coords.current = {
      start: { x: offset.left, y: offset.top },
      offset: { x, y }
    };

    document.addEventListener('mousemove', drag);
    document.addEventListener('mouseup', dragEnd);
    document.addEventListener('touchmove', drag, { passive: false });
    document.addEventListener('touchend', dragEnd);
    document.addEventListener('touchcancel', dragEnd);
  };

  return (
    <div
      {...themer('xyControlContainer')}
      data-testid="xy"
      ref={xyControlContainer}
      onTouchStart={dragStart}
      onMouseDown={dragStart}
    >
      <div
        {...themer('xyControl', `${isDark ? 'xyControlDark' : ''}`)}
        style={{
          backgroundColor,
          width: `${discRadius}px`,
          height: `${discRadius}px`,
          top: `calc(${top}% - ${discRadius / 2}px)`,
          left: `calc(${left}% - ${discRadius / 2}px)`
        }}
      />
      {children}
    </div>
  );
}

export { XYInput };
