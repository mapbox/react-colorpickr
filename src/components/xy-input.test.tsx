import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import { XYInput } from './xy-input';

describe('XYInput', () => {
  const props = {
    theme: {},
    x: 0,
    y: 10,
    xmax: 100,
    ymax: 100,
    isDark: false,
    discRadius: 18,
    onChange: jest.fn()
  };

  test('renders', () => {
    const { getByTestId, baseElement } = render(
      <XYInput {...props}>
        <span>children</span>
      </XYInput>
    );
    expect(baseElement).toMatchSnapshot();
    const xy = getByTestId('xy');

    fireEvent.mouseMove(xy, { clientX: 10, clientY: 10 });
    fireEvent.touchMove(xy, {
      changedTouches: [
        {
          clientX: 10,
          clientY: 10
        }
      ]
    });
    expect(props.onChange).toHaveBeenCalledTimes(0);

    fireEvent.mouseDown(xy, { clientX: 0, clientY: 0 });
    expect(props.onChange).toHaveBeenCalledTimes(1);
    fireEvent.touchStart(xy, {
      changedTouches: [
        {
          clientX: 0,
          clientY: 0
        }
      ]
    });
    expect(props.onChange).toHaveBeenCalledTimes(2);
    fireEvent.mouseMove(xy, { clientX: 10, clientY: 10 });
    expect(props.onChange).toHaveBeenCalledTimes(3);
    fireEvent.touchMove(xy, {
      changedTouches: [
        {
          clientX: 10,
          clientY: 10
        }
      ]
    });
    expect(props.onChange).toHaveBeenCalledTimes(4);

    fireEvent.mouseUp(xy);
    fireEvent.touchEnd(xy);
    fireEvent.mouseMove(xy, { clientX: 10, clientY: 10 });
    fireEvent.touchMove(xy, {
      changedTouches: [
        {
          clientX: 10,
          clientY: 10
        }
      ]
    });
    expect(props.onChange).toHaveBeenCalledTimes(4);
  });

  test('renders isDark', () => {
    const isDarkProps = {
      ...props,
      isDark: true
    };

    const { baseElement } = render(
      <XYInput {...isDarkProps}>
        <span>children</span>
      </XYInput>
    );
    expect(baseElement).toMatchSnapshot();
  });
});
