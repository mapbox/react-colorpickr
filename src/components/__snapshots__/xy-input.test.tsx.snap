// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`XYInput renders 1`] = `
<body>
  <div>
    <div
      data-testid="xy"
    >
      <div
        style="width: 18px; height: 18px; top: calc(10% - 9px); left: calc(0% - 9px);"
      />
      <span>
        children
      </span>
    </div>
  </div>
</body>
`;

exports[`XYInput renders isDark 1`] = `
<body>
  <div>
    <div
      data-testid="xy"
    >
      <div
        style="width: 18px; height: 18px; top: calc(10% - 9px); left: calc(0% - 9px);"
      />
      <span>
        children
      </span>
    </div>
  </div>
</body>
`;
